pvanalytics/features/clipping.py

```diff
@@ -225,3 +225,178 @@ def threshold(ac_power, slope_max=0.0035, power_min=0.75,
+    daily_clipped_min = daily_mean - 2 * daily_std
+    return daily_clipped_min, daily_clipped_max
+
+
+def _threshold_minmax(mask, data):
+    """Return daily thresholds based on min and max.
+
+    Parameters
+    ----------
+    mask : Series
+        Boolean series
+    data : Series
+        Data with same index as `mask`.
+
+    Returns
+    -------
+    minimum : Series
+        `data` transformed to have the minimum value from ``data[mask]``
+        on each day.
+    maximum : Series
+        `data` transformed to have the maximum value from ``data[mask]``
+        on each day.
+    """
+    daily_max = _apply_daily_mask(mask, data, pd.Series.max)
+    daily_min = _apply_daily_mask(mask, data, pd.Series.min)
+    return daily_min, daily_max
+
+
+def _rolling_low_slope(ac_power, window, slope_min):
+    """Return True for timestamps where the data has slope less
+    than `slope_min` over a rolling window of length `window."""
+    rolling_max = ac_power.rolling(window=window).max()
+    rolling_min = ac_power.rolling(window=window).min()
+    # calculate an upper bound on the derivative
+    derivative_max = ((rolling_max - rolling_min)
+                      / ((rolling_max + rolling_min) / 2) * 100)
+    clipped = derivative_max < slope_min
+    clipped_windows = clipped.copy()
+    # flag all points in a window that has clipping
+    for i in range(0, window):
+        clipped_windows |= clipped.shift(-i)
+    return clipped_windows
+
+
+def geometric(ac_power, window=None, slope_min=0.2, freq=None,
+              tracking=False):
+    """Identify clipping based on a the shape of the `ac_power`
+    curve on each day.
+
+    Each day is checked for periods where the slope of `ac_power`
+    is low. The power values in these times are used to calculate
+    a minimum and a maximum clipped power level for that day. Any
+    power values that are within this range are flagged as
+    clipped. The methodology for computing the thresholds varies
+    depending on the frequency of `ac_power`. For high frequency
+    data, less than 10 minute timestamp spacing the minimum
+    clipped power is the mean of the low-slope period(s) on that
+    day minus 2 times the standard deviation in the same period(s).
+    For higher frequency data the absolute minimum and maximum of
+    the low slope period(s) on each day are used.
+
+    If the frequency of `ac_power` is less than ten minutes, then
+    `ac_power` is down-sampled to 15 minutes and the mean value in
+    each 15-minute period is used to reduce noise inherent in
+    high frequency data.
+
+    Parameters
+    ----------
+    ac_power : Series
+        AC power data.
+    window : int, optional
+        Size of the rolling window used to identify low-slope
+        periods. If not specified and `tracking` is False then
+        `window=3` is used. If not specified and `tracking` is
+        True then `window=5` is used.
+    slope_min : float, default 0.2
```

Comment:
Does this have a unit?  AC power / minute?

Reply:
I'm struggling a little to describe this. It is the percent change per window.

Reply:
Yeah.

```
    derivative_max = ((rolling_max - rolling_min)
                      / ((rolling_max + rolling_min) / 2) * 100)
    clipped = derivative_max < slope_min
```

Suggest "Maximum difference in maximum and minimum power for a window to be flagged as clipped. Units are percent of average power in the interval."

I think this is a maximum not a minimum slope, maybe `slope_max`